fe/fe-core/src/main/java/com/starrocks/meta/lock/MultiUserLock.java

```diff
@@ -0,0 +1,351 @@
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package com.starrocks.meta.lock;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+
+public class MultiUserLock extends Lock {
+    protected LockHolder firstOwner;
+    protected LockHolder firstWaiter;
+    private List<LockHolder> otherOwners;
+    private List<LockHolder> otherWaiters;
+
+    public MultiUserLock(LockHolder lockHolder) {
+        this.firstOwner = lockHolder;
+    }
+
+    @Override
+    public LockGrantType lock(Locker locker, LockType lockType) {
+        LockHolder lockHolderRequest = new LockHolder(locker, lockType);
+        LockGrantType lockGrantType = tryLock(lockHolderRequest);
+        if (lockGrantType == LockGrantType.NEW) {
+            addOwner(lockHolderRequest);
+        } else if (lockGrantType == LockGrantType.WAIT) {
+            addWaiterToEnd(lockHolderRequest);
+        }
+
+        return lockGrantType;
+    }
+
+    private LockGrantType tryLock(LockHolder lockHolderRequest) {
+        if (ownerNum() == 0) {
+            return LockGrantType.NEW;
+        }
+
+        boolean hasConflicts = false;
+
+        LockHolder lockOwner = null;
+        Iterator<LockHolder> ownerIterator = null;
+        if (otherOwners != null) {
+            ownerIterator = otherOwners.iterator();
+        }
+
+        if (firstOwner != null) {
+            lockOwner = firstOwner;
+        } else {
+            if (ownerIterator != null && ownerIterator.hasNext()) {
+                lockOwner = ownerIterator.next();
+            }
+        }
+
+        LockHolder sameLockHolder = null;
+        while (lockOwner != null) {
+            if (lockHolderRequest.equals(lockOwner)) {
+                sameLockHolder = lockOwner;
+            } else {
+                boolean isConflict = lockOwner.isConflict(lockHolderRequest);
+                if (isConflict) {
+                    hasConflicts = true;
+                }
+            }
+
+            if (ownerIterator != null && ownerIterator.hasNext()) {
+                lockOwner = ownerIterator.next();
+            } else {
+                lockOwner = null;
+            }
+        }
+
+        if (hasConflicts) {
+            return LockGrantType.WAIT;
+        } else {
+            if (sameLockHolder != null) {
+                sameLockHolder.increaseRefCount();
+                return LockGrantType.EXISTING;
+            } else {
+                if (waiterNum() == 0) {
+                    return LockGrantType.NEW;
+                } else {
+                    return LockGrantType.WAIT;
```

Comment:
add comments to clarify that this's FAIR strategy, if there's no conflict, but has pending waiters, the after lock req has to wait.